src/java.base/share/classes/java/util/zip/GZIPOutputStream.java

```diff
@@ -179,21 +255,118 @@ public void finish() throws IOException {
+            flags |= FNAME;
+        }
+        if (fileComment != null && fileComment.length != 0) {
+            flags |= FCOMMENT;
+        }
+
+        // the head of header.
+        byte [] head = new byte[] {
+                           (byte) GZIP_MAGIC,        // Magic number (short)
+                           (byte)(GZIP_MAGIC >> 8),  // Magic number (short)
+                           Deflater.DEFLATED,        // Compression method (CM)
+                           flags,                    // Flags (FLG)
+                           0,                        // Modification time MTIME (int)
+                           0,                        // Modification time MTIME (int)
+                           0,                        // Modification time MTIME (int)
+                           0,                        // Modification time MTIME (int)
+                           0,                        // Extra flags (XFLG)
+                           OS_UNKNOWN                // Operating system (OS)
+        };
+        // write head.
+        out.write(head);
+        if (generateHeaderCRC) {
+            crc.update(head, 0, head.length);
+        }
+
+        // write extra field.
+        if ((flags & FEXTRA) == FEXTRA) {
+            /* extra field, per RFC-1952
+             *     +---+---+=================================+
+             *     | XLEN  |...XLEN bytes of "extra field"...|
+             *     +---+---+=================================+
+             */
+            int xlen = extraFieldBytes.length;
+            if (xlen > 0xffff) {
+                throw new ZipException("extra field size out of range");
+            }
+            // write XLEN.
+            out.write((byte)(xlen & 0xff));
+            out.write((byte)((xlen >> 8) & 0xff));
+            // write extra field bytes.
+            out.write(extraFieldBytes);
+            if (generateHeaderCRC) {
+                crc.update((byte)(xlen & 0xff));
+                crc.update(((byte)(xlen >> 8) & 0xff));
+                crc.update(extraFieldBytes, 0, extraFieldBytes.length);
+            }
+        }
+        // write file name.
+        if ((flags & FNAME) == FNAME) {
+            /*
+             *    +=========================================+
+             *    |...original file name, zero-terminated...|
+             *    +=========================================+
+             */
+            out.write(filename);
+            out.write(0);
+            if (generateHeaderCRC) {
+                crc.update(filename, 0, filename.length);
+                crc.update(0);
+            }
+        }
+        // write file comment.
+        if ((flags & FCOMMENT) == FCOMMENT) {
+            /*
+             *    +===================================+
+             *    |...file comment, zero-terminated...|
+             *    +===================================+
+             */
+            out.write(fileComment);
+            out.write(0);
```

Comment:
Same as file name , I think it is the users responsibility to pass the right encoded String as an argument. if constructor is designed to accept a String here, I think it can decode to byte array with ISO 8859-1, and then write to the header.

Moreover，I can't figure out a way to testing the encoding information of the String, do you have any clue about validate it?